File: frontend/src/types/execution.ts
/**
 * Execution records, trace events, and hierarchical trace tree models.
 */

import { PaginatedResponse } from './api';

export type ExecutionStatus = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED' | 'TIMEOUT';

export interface ExecutionSummary {
  id: string;
  workflow_definition_id: string;
  started_at: string;
  completed_at?: string | null;
  duration_ms: number;
  status: ExecutionStatus;
  retry_count: number;
  error_count: number;
  failure_reason?: string | null;
  incident_id?: string | null;
  is_incident_affected: boolean;
  metadata?: Record<string, unknown>;
}

export type ExecutionListResponse = PaginatedResponse<ExecutionSummary>;

export type EventType =
  | 'WORKFLOW_START'
  | 'WORKFLOW_COMPLETE'
  | 'WORKFLOW_FAIL'
  | 'SPAN_START'
  | 'SPAN_COMPLETE'
  | 'SPAN_ERROR'
  | 'SPAN_TIMEOUT'
  | 'RETRY_ATTEMPT';

export interface TraceEvent {
  event_id: string;
  timestamp: string;
  execution_id: string;
  workflow_id: string;
  service: string;
  operation: string;
  event_type: EventType;
  status: string;
  latency_ms: number;
  parent_event_id?: string | null;
  correlation_id?: string | null;
  metadata?: Record<string, unknown>;
}

export interface TraceTreeNode {
  event_id: string;
  timestamp: string;
  execution_id: string;
  workflow_id: string;
  service: string;
  operation: string;
  event_type: EventType;
  status: string;
  latency_ms: number;
  parent_event_id?: string | null;
  correlation_id?: string | null;
  metadata?: Record<string, unknown>;
  children: TraceTreeNode[];
}

File: apps/api/schemas/execution.py
"""Pydantic v2 schemas for workflow executions, trace events, and DAG trees."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from apps.api.schemas.common import PaginationMeta


class ExecutionSummaryResponse(BaseModel):
    """Workflow execution record summary response."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    workflow_definition_id: str
    started_at: datetime
    completed_at: datetime | None
    duration_ms: float
    status: str
    retry_count: int
    error_count: int
    failure_reason: str | None
    incident_id: str | None
    is_incident_affected: bool
    metadata: dict[str, Any]


class TraceEventResponse(BaseModel):
    """Individual span lifecycle event schema."""

    model_config = ConfigDict(from_attributes=True)

    event_id: str
    timestamp: datetime
    execution_id: str
    workflow_id: str
    service: str
    operation: str
    event_type: str
    status: str
    latency_ms: float
    parent_event_id: str | None
    correlation_id: str | None
    metadata: dict[str, Any]


class TraceTreeNodeResponse(BaseModel):
    """Reconstructed recursive DAG tree node with child spans."""

    event_id: str
    service: str
    operation: str
    event_type: str
    status: str
    latency_ms: float
    timestamp: str
    metadata: dict[str, Any] = Field(default_factory=dict)
    children: list["TraceTreeNodeResponse"] = Field(default_factory=list)


class ExecutionListResponse(BaseModel):
    """Paginated list of workflow executions."""

    items: list[ExecutionSummaryResponse]
    pagination: PaginationMeta
